File: skills/code-security-review/scripts/lib/report.py
#!/usr/bin/env python3
"""Aggregate findings.jsonl into a human-readable Markdown report.

Usage:
    python3 lib/report.py out/findings.jsonl > out/report.md
    python3 lib/report.py out/findings.jsonl --json > out/summary.json
"""

# Findings are JSON records with intentionally heterogeneous evidence shapes; Any is deliberate.
# pyright: reportAny=false, reportExplicitAny=false, reportUnknownMemberType=false, reportUnknownArgumentType=false, reportUnknownVariableType=false

from __future__ import annotations

import json
import sys
from collections import Counter, defaultdict
from pathlib import Path
from typing import Any

SEVERITY_ORDER: tuple[str, ...] = ("critical", "high", "medium", "low", "info")
SEVERITY_BADGE: dict[str, str] = {
    "critical": "[CRITICAL]",
    "high":     "[HIGH]",
    "medium":   "[MEDIUM]",
    "low":      "[LOW]",
    "info":     "[INFO]",
}

Finding = dict[str, Any]


def load_findings(path: Path) -> list[Finding]:
    findings: list[Finding] = []
    if not path.exists() or path.stat().st_size == 0:
        return findings
    for lineno, line in enumerate(path.read_text(encoding="utf-8").splitlines(), 1):
        line = line.strip()
        if not line:
            continue
        try:
            findings.append(json.loads(line))
        except json.JSONDecodeError as e:
            print(f"# warning: skipping malformed line {lineno}: {e}", file=sys.stderr)
    return findings


def render_markdown(findings: list[Finding]) -> str:
    counts: Counter[str] = Counter(str(f.get("severity", "info")) for f in findings)
    by_phase: dict[str, list[Finding]] = defaultdict(list)
    for f in findings:
        by_phase[str(f.get("phase", "unknown"))].append(f)

    lines: list[str] = []
    lines.append("# Code Security Review — Active Probe Report")
    lines.append("")
    lines.append(f"_Generated from {len(findings)} finding(s)._")
    lines.append("")

    # --- Summary table ---
    lines.append("## Summary")
    lines.append("")
    lines.append("| Severity | Count |")
    lines.append("| --- | --- |")
    for sev in SEVERITY_ORDER:
        lines.append(f"| {SEVERITY_BADGE[sev]} | {counts.get(sev, 0)} |")
    lines.append(f"| **Total** | **{len(findings)}** |")
    lines.append("")

    if not findings:
        lines.append("> No findings recorded. Either the API passed all probes or no scripts were run.")
        lines.append("")
        return "\n".join(lines)

    # --- Verdict ---
    if counts.get("critical", 0) > 0:
        verdict = "FAIL — critical findings present, do not deploy"
    elif counts.get("high", 0) > 0:
        verdict = "FAIL — high-severity findings require remediation before deploy"
    elif counts.get("medium", 0) > 0:
        verdict = "REVIEW — medium-severity findings require triage"
    elif counts.get("low", 0) > 0:
        verdict = "PASS WITH NOTES — only low-severity findings"
    else:
        verdict = "PASS — only informational findings"
    lines.append(f"**Verdict:** {verdict}")
    lines.append("")

    # --- Findings grouped by severity ---
    for sev in SEVERITY_ORDER:
        sev_findings = [f for f in findings if f.get("severity") == sev]
        if not sev_findings:
            continue
        lines.append(f"## {SEVERITY_BADGE[sev]} ({len(sev_findings)})")
        lines.append("")
        for f in sev_findings:
            lines.append(f"### `{f.get('id', '?')}` — {f.get('endpoint', '?')}")
            lines.append("")
            lines.append(f"- **Phase:** `{f.get('phase', '?')}`")
            lines.append(f"- **OWASP:** {f.get('owasp', '—')}")
            lines.append(f"- **Timestamp:** {f.get('timestamp', '—')}")
            lines.append(f"- **Impact:** {f.get('impact', '—')}")
            lines.append(f"- **Remediation:** {f.get('remediation', '—')}")
            evidence = f.get("evidence") or {}
            if evidence:
                lines.append("")
                lines.append("```json")
                lines.append(json.dumps(evidence, indent=2, sort_keys=True))
                lines.append("```")
            lines.append("")

    # --- Findings grouped by phase (compact) ---
    lines.append("## By Phase")
    lines.append("")
    lines.append("| Phase | Findings | Highest severity |")
    lines.append("| --- | --- | --- |")
    for phase, items in sorted(by_phase.items()):
        sevs = {item.get("severity") for item in items}
        for sev in SEVERITY_ORDER:
            if sev in sevs:
                highest = sev.upper()
                break
        else:
            highest = "—"
        lines.append(f"| `{phase}` | {len(items)} | {highest} |")
    lines.append("")

    return "\n".join(lines)


def render_json_summary(findings: list[Finding]) -> str:
    counts: Counter[str] = Counter(str(f.get("severity", "info")) for f in findings)
    summary: dict[str, Any] = {
        "total":  len(findings),
        "counts": {sev: counts.get(sev, 0) for sev in SEVERITY_ORDER},
        "verdict_fail": counts.get("critical", 0) > 0 or counts.get("high", 0) > 0,
    }
    return json.dumps(summary, indent=2)


def main(argv: list[str]) -> int:
    if len(argv) < 2:
        print(__doc__, file=sys.stderr)
        return 1
    path = Path(argv[1])
    findings = load_findings(path)
    if "--json" in argv[2:]:
        print(render_json_summary(findings))
    else:
        print(render_markdown(findings))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: skills/context-optimization/scripts/ctxopt/src/report.rs
//! `ctxopt report` — aggregate findings into Markdown report + summary {JSON, TOON}.
//! Equivalent to legacy lib/report.py.

use crate::cli::ReportFormat;
use crate::finding::{Category, Finding, Severity};
use crate::toon;
use anyhow::{anyhow, Context, Result};
use std::collections::HashMap;
use std::fs::File;
use std::io::{BufRead, BufReader, BufWriter, Write};
use std::path::Path;

const SEVERITY_ORDER: [Severity; 5] = [
    Severity::Critical,
    Severity::High,
    Severity::Medium,
    Severity::Low,
    Severity::Info,
];

const CATEGORY_LABELS: &[(Category, &str)] = &[
    (Category::Cost, "Cost waste"),
    (Category::Cache, "Cache miss"),
    (Category::Quality, "Quality degradation"),
    (Category::Bloat, "Token bloat"),
    (Category::Architecture, "Architecture risk"),
];

fn category_label(c: Category) -> &'static str {
    CATEGORY_LABELS
        .iter()
        .find(|(cat, _)| *cat == c)
        .map(|(_, l)| *l)
        .unwrap_or("Unknown")
}

fn severity_badge(s: Severity) -> &'static str {
    match s {
        Severity::Critical => "[CRITICAL]",
        Severity::High => "[HIGH]",
        Severity::Medium => "[MEDIUM]",
        Severity::Low => "[LOW]",
        Severity::Info => "[INFO]",
    }
}

/// Detect input format by extension and parse into a Vec<Finding>.
pub fn load_findings(path: &Path) -> Result<Vec<Finding>> {
    let ext = path
        .extension()
        .and_then(|e| e.to_str())
        .unwrap_or("")
        .to_lowercase();

    match ext.as_str() {
        "jsonl" => load_jsonl(path),
        "json" => load_json(path),
        "toon" => Err(anyhow!(
            "TOON parser not implemented for input. Use --format=jsonl or --format=json source."
        )),
        _ => {
            // Best-effort: try jsonl first, fall back to json.
            load_jsonl(path).or_else(|_| load_json(path))
        }
    }
}

fn load_jsonl(path: &Path) -> Result<Vec<Finding>> {
    let f = File::open(path).with_context(|| format!("opening {}", path.display()))?;
    let reader = BufReader::new(f);
    let mut out = Vec::new();
    for (lineno, line) in reader.lines().enumerate() {
        let line = match line {
            Ok(l) => l,
            Err(_) => continue,
        };
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        match serde_json::from_str::<Finding>(line) {
            Ok(f) => out.push(f),
            Err(e) => eprintln!(
                "# warning: skipping malformed line {} in {}: {}",
                lineno + 1,
                path.display(),
                e
            ),
        }
    }
    Ok(out)
}

fn load_json(path: &Path) -> Result<Vec<Finding>> {
    let text = std::fs::read_to_string(path)
        .with_context(|| format!("reading {}", path.display()))?;
    let v: Vec<Finding> = serde_json::from_str(&text)
        .with_context(|| format!("parsing JSON array from {}", path.display()))?;
    Ok(v)
}

/// Render the standard Markdown report.
pub fn render_markdown(findings: &[Finding]) -> String {
    let mut counts: HashMap<Severity, usize> = HashMap::new();
    let mut by_phase: HashMap<String, Vec<&Finding>> = HashMap::new();
    let mut by_category: HashMap<Category, usize> = HashMap::new();

    for f in findings {
        *counts.entry(f.severity).or_insert(0) += 1;
        by_phase.entry(f.phase.clone()).or_default().push(f);
        *by_category.entry(f.category).or_insert(0) += 1;
    }

    let total = findings.len();
    let mut lines: Vec<String> = Vec::new();
    lines.push("# Context Optimization — Audit Report".into());
    lines.push(String::new());
    lines.push(format!("_Generated from {total} finding(s)._"));
    lines.push(String::new());

    // Summary by severity
    lines.push("## Summary by severity".into());
    lines.push(String::new());
    lines.push("| Severity | Count |".into());
    lines.push("| --- | --- |".into());
    for sev in SEVERITY_ORDER {
        lines.push(format!(
            "| {} | {} |",
            severity_badge(sev),
            counts.get(&sev).copied().unwrap_or(0)
        ));
    }
    lines.push(format!("| **Total** | **{total}** |"));
    lines.push(String::new());

    if findings.is_empty() {
        lines.push(
            "> No findings recorded. Either the context passes the audit or no scripts were run."
                .into(),
        );
        lines.push(String::new());
        return lines.join("\n");
    }

    // Verdict
    let crit = counts.get(&Severity::Critical).copied().unwrap_or(0);
    let high = counts.get(&Severity::High).copied().unwrap_or(0);
    let med = counts.get(&Severity::Medium).copied().unwrap_or(0);
    let low = counts.get(&Severity::Low).copied().unwrap_or(0);
    let verdict = if crit > 0 {
        "REWORK — critical findings present, optimization plan needs immediate adoption"
    } else if high > 0 {
        "REVIEW — high-severity findings require targeted remediation"
    } else if med > 0 {
        "TUNE — medium-severity findings warrant attention"
    } else if low > 0 {
        "PASS WITH NOTES — only low-severity findings"
    } else {
        "PASS — only informational findings"
    };
    lines.push(format!("**Verdict:** {verdict}"));
    lines.push(String::new());

    // Summary by category
    lines.push("## Summary by category".into());
    lines.push(String::new());
    lines.push("| Category | Findings |".into());
    lines.push("| --- | --- |".into());
    let mut cat_keys: Vec<Category> = by_category.keys().copied().collect();
    cat_keys.sort_by_key(|c| category_label(*c));
    for cat in cat_keys {
        lines.push(format!(
            "| {} | {} |",
            category_label(cat),
            by_category[&cat]
        ));
    }
    lines.push(String::new());

    // Findings grouped by severity
    for sev in SEVERITY_ORDER {
        let sev_findings: Vec<&Finding> = findings.iter().filter(|f| f.severity == sev).collect();
        if sev_findings.is_empty() {
            continue;
        }
        lines.push(format!("## {} ({})", severity_badge(sev), sev_findings.len()));
        lines.push(String::new());
        for f in sev_findings {
            lines.push(format!("### `{}` — {}", f.id, f.target));
            lines.push(String::new());
            lines.push(format!("- **Phase:** `{}`", f.phase));
            lines.push(format!("- **Category:** {}", category_label(f.category)));
            lines.push(format!("- **Timestamp:** {}", f.timestamp));
            lines.push(format!("- **Impact:** {}", f.impact));
            lines.push(format!("- **Remediation:** {}", f.remediation));
            if !f.evidence.is_null() && f.evidence != serde_json::json!({}) {
                lines.push(String::new());
                lines.push("```json".into());
                lines.push(serde_json::to_string_pretty(&f.evidence).unwrap_or_default());
                lines.push("```".into());
            }
            lines.push(String::new());
        }
    }

    // By phase
    lines.push("## By phase".into());
    lines.push(String::new());
    lines.push("| Phase | Findings | Highest severity |".into());
    lines.push("| --- | --- | --- |".into());
    let mut phase_keys: Vec<String> = by_phase.keys().cloned().collect();
    phase_keys.sort();
    for phase in &phase_keys {
        let items = &by_phase[phase];
        let highest = SEVERITY_ORDER
            .iter()
            .find(|sev| items.iter().any(|f| f.severity == **sev))
            .map(|s| s.as_str().to_uppercase())
            .unwrap_or_else(|| "—".to_string());
        lines.push(format!("| `{phase}` | {} | {highest} |", items.len()));
    }
    lines.push(String::new());

    lines.join("\n")
}

/// Render the JSON summary used by CI / aggregators.
pub fn render_json_summary(findings: &[Finding]) -> String {
    let mut counts: HashMap<Severity, usize> = HashMap::new();
    let mut by_category: HashMap<Category, usize> = HashMap::new();
    for f in findings {
        *counts.entry(f.severity).or_insert(0) += 1;
        *by_category.entry(f.category).or_insert(0) += 1;
    }
    let summary = serde_json::json!({
        "total": findings.len(),
        "counts": {
            "critical": counts.get(&Severity::Critical).copied().unwrap_or(0),
            "high": counts.get(&Severity::High).copied().unwrap_or(0),
            "medium": counts.get(&Severity::Medium).copied().unwrap_or(0),
            "low": counts.get(&Severity::Low).copied().unwrap_or(0),
            "info": counts.get(&Severity::Info).copied().unwrap_or(0),
        },
        "by_category": {
            "cost": by_category.get(&Category::Cost).copied().unwrap_or(0),
            "cache": by_category.get(&Category::Cache).copied().unwrap_or(0),
            "quality": by_category.get(&Category::Quality).copied().unwrap_or(0),
            "bloat": by_category.get(&Category::Bloat).copied().unwrap_or(0),
            "architecture": by_category.get(&Category::Architecture).copied().unwrap_or(0),
        },
        "verdict_rework": counts.get(&Severity::Critical).copied().unwrap_or(0) > 0,
        "verdict_review": counts.get(&Severity::High).copied().unwrap_or(0) > 0,
    });
    serde_json::to_string_pretty(&summary).unwrap_or_default()
}

/// Render the TOON summary block.
pub fn render_toon_summary(findings: &[Finding]) -> String {
    let mut counts: HashMap<Severity, usize> = HashMap::new();
    let mut by_category: HashMap<Category, usize> = HashMap::new();
    for f in findings {
        *counts.entry(f.severity).or_insert(0) += 1;
        *by_category.entry(f.category).or_insert(0) += 1;
    }
    let mut buf: Vec<u8> = Vec::new();
    let severity_counts: Vec<(&str, usize)> = SEVERITY_ORDER
        .iter()
        .map(|s| (s.as_str(), counts.get(s).copied().unwrap_or(0)))
        .collect();
    let category_counts: Vec<(&str, usize)> = CATEGORY_LABELS
        .iter()
        .map(|(c, _)| (c.as_str(), by_category.get(c).copied().unwrap_or(0)))
        .collect();
    let verdict_rework = counts.get(&Severity::Critical).copied().unwrap_or(0) > 0;
    let verdict_review = counts.get(&Severity::High).copied().unwrap_or(0) > 0;
    let verdict_pass_with_notes = counts.get(&Severity::Low).copied().unwrap_or(0) > 0
        && counts.get(&Severity::Critical).copied().unwrap_or(0) == 0
        && counts.get(&Severity::High).copied().unwrap_or(0) == 0
        && counts.get(&Severity::Medium).copied().unwrap_or(0) == 0;
    toon::write_summary(
        &mut buf,
        findings.len(),
        &severity_counts,
        &category_counts,
        verdict_rework,
        verdict_review,
        verdict_pass_with_notes,
    )
    .ok();
    String::from_utf8(buf).unwrap_or_default()
}

/// Top-level driver for the `report` subcommand.
pub fn run(findings_path: &Path, format: ReportFormat) -> Result<()> {
    let findings = load_findings(findings_path)?;
    let stdout = std::io::stdout();
    let mut handle = stdout.lock();
    match format {
        ReportFormat::Md => handle.write_all(render_markdown(&findings).as_bytes())?,
        ReportFormat::Json => handle.write_all(render_json_summary(&findings).as_bytes())?,
        ReportFormat::Toon => handle.write_all(render_toon_summary(&findings).as_bytes())?,
    }
    handle.write_all(b"\n")?;
    Ok(())
}

/// Helper: write a report file (used by run-all to bundle outputs).
pub fn write_report_files(out_dir: &Path, findings: &[Finding]) -> Result<()> {
    std::fs::create_dir_all(out_dir)?;

    let report_md = out_dir.join("report.md");
    let mut f = BufWriter::new(File::create(&report_md)?);
    f.write_all(render_markdown(findings).as_bytes())?;
    f.write_all(b"\n")?;

    let summary_json = out_dir.join("summary.json");
    let mut f = BufWriter::new(File::create(&summary_json)?);
    f.write_all(render_json_summary(findings).as_bytes())?;
    f.write_all(b"\n")?;

    let summary_toon = out_dir.join("summary.toon");
    let mut f = BufWriter::new(File::create(&summary_toon)?);
    f.write_all(render_toon_summary(findings).as_bytes())?;

    Ok(())
}
